### PR TITLE
update to 1.12

### PR DESCRIPTION
**Pull request by DerBaM**
My version of an update to 1.12. I know that gudenau allready poste one but he changed a lot unneaded (in my opinion) and that he don't checked his commits (my IDE done this etc.) make this not better.

So my try:
The only big change is the configuration loading. The AbstractConfig.load() would not work (MC crash) so I decided to fix this with "copying" some code from Tinkers Construct istead of fix Mantle to only commit in one place.

In single player tested:
- level up of tools
- level up of weapons
- level up ding and meassage
- config file changes have affect

Fixed also all warnings of my IDE (eclipse) except the deprecated I18n ones.

### CODE REVIEW

**Comment by boosterdev** [comment]
for server this does not work
java.lang.NoClassDefFoundError: net/minecraft/client/util/ITooltipFlag

**Comment by DerBaM** [comment]
Fixed this and the config comments. I've changed them to be clearer but make mistakes.

On server tested:
- level up of tools
- level up of weapons
- level up meassage
- config file changes have affect

Not working on server (for me) is the level up ding.

**Comment by boosterdev** [comment]
nice work 👍

File: src/main/java/slimeknights/toolleveling/EntityXpHandler.java [hunk 1]
```diff
@@ -15,16 +16,18 @@
...
+    if (event.getObject() instanceof EntityLivingBase && event.getObject().isEntityAlive()) {
       event.addCapability(CAPABILITY_KEY, new DamageXpHandler());
     }
   }
 
+  @SuppressWarnings({"ConstantConditions", "unused"})
   @SubscribeEvent
   public void onDeath(LivingDeathEvent event) {
-    if(!event.getEntity().getEntityWorld().isRemote && event.getEntity().hasCapability(CapabilityDamageXp.CAPABILITY, null)) {
+    if (!event.getEntity().getEntityWorld().isRemote && event.getEntity().hasCapability(CapabilityDamageXp.CAPABILITY, null)) {
```
Comment (KnightMiner) [inline src/main/java/slimeknights/toolleveling/EntityXpHandler.java]: This is minor, but typically we don't have spaces after "if"

File: src/main/java/slimeknights/toolleveling/Tooltips.java [hunk 2]
```diff
@@ -1,99 +1,49 @@
...
+import net.minecraftforge.fml.relauncher.Side;
+import net.minecraftforge.fml.relauncher.SideOnly;
 import slimeknights.tconstruct.library.utils.TinkerUtil;
 import slimeknights.toolleveling.config.Config;
 
+import java.util.List;
+
 // utility class for constructing tooltip
-public final class Tooltips {
+@SideOnly(Side.CLIENT)
```
Comment (KnightMiner) [inline src/main/java/slimeknights/toolleveling/Tooltips.java]: Typically @SideOnly should not be applied to a class as it was never more than a hack. Is there a way to just not reference this class in serverside code like is done with client proxies?

File: src/main/resources/assets/tinkertoolleveling/lang/en_us.lang [hunk 3]
```diff
@@ -1,31 +1,4 @@
 tooltip.level=Level
 tooltip.xp=XP
 
-tooltip.level.0=Like new
```
Comment (KnightMiner) [inline src/main/resources/assets/tinkertoolleveling/lang/en_us.lang]: Why are these removed? Sure, it does not support every level, but it was much nicer flavor text

File: src/main/resources/assets/tinkertoolleveling/lang/de_de.lang [hunk 4]
```diff
@@ -0,0 +1,4 @@
+tooltip.level=Stufe
```
Comment (KnightMiner) [inline src/main/resources/assets/tinkertoolleveling/lang/de_de.lang]: While nice to have, I feel this is better as a separate PR to keep one topic per PR. I'd double check with boni though.

File: src/main/resources/assets/tinkertoolleveling/lang/en_us.lang (hunk 3 above)
Comment (DerBaM) [inline src/main/resources/assets/tinkertoolleveling/lang/en_us.lang]: It cycled so e.g. clumsy+ is better then awesome. And the color cycle even more:
level * 0.2666 % 1
so there is no info out of the color or the text. Okey, on the other side how much level will you get during playthrough. If this is a deal I have an idea to make us both happy, if you want.

File: src/main/resources/assets/tinkertoolleveling/lang/de_de.lang (hunk 4 above)
Comment (DerBaM) [inline src/main/resources/assets/tinkertoolleveling/lang/de_de.lang]: I was in flow (and had time) so I've made my way. But I'm actually all is done what I wanted to finish. So no more comits.

File: src/main/java/slimeknights/toolleveling/EntityXpHandler.java (hunk 1 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/EntityXpHandler.java]: Change my IDE settings for this project and fixed all.

File: src/main/java/slimeknights/toolleveling/ModToolLeveling.java [hunk 5]
```diff
@@ -93,36 +93,37 @@ public void onBlock(ItemStack tool, EntityPlayer player, LivingHurtEvent event)
...
+  @SuppressWarnings("unused")
   @SubscribeEvent
   public void onPath(TinkerToolEvent.OnShovelMakePath event) {
     addXp(event.itemStack, 1, event.player);
   }
 
+  @SuppressWarnings("unused")
   @SubscribeEvent(priority = EventPriority.LOWEST, receiveCanceled = true)
   public void onLivingHurt(LivingAttackEvent event) {
-    // if it's cancelled it got handled by the battlesign (or something else. but it's a prerequisite.)
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/ModToolLeveling.java]: Those ifs were split up for readability.

File: src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java [hunk 6]
```diff
@@ -9,38 +9,32 @@
...
 import slimeknights.toolleveling.capability.CapabilityDamageXp;
 import slimeknights.toolleveling.config.Config;
 import slimeknights.toolleveling.config.ConfigSync;
 import slimeknights.toolleveling.config.ConfigSyncPacket;
-import slimeknights.toolleveling.debug.CommandLevelTool;
 
+@SuppressWarnings("WeakerAccess")
 @Mod(modid = TinkerToolLeveling.MODID,
-    version = TinkerToolLeveling.VERSION,
-    name = "TinkerToolLeveling",
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java]: Metadata is missing

File: src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java [hunk 7]
```diff
@@ -52,20 +46,25 @@ public void preInit(FMLPreInitializationEvent event) {
     Config.INSTANCE.save();
   }
 
+  @SuppressWarnings("unused")
   @EventHandler
   public void init(FMLInitializationEvent event) {
+    /* Need this here, because preInit doesn't guarantee that TTL will
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java]: It does. You removed the dependency that ensures exactly that.

File: src/main/java/slimeknights/toolleveling/CommonProxy.java [hunk 8]
```diff
@@ -4,38 +4,53 @@
...
+        "message.levelup",
+        itemStack.getDisplayName(),
+        componentLvl
+    );
+    componentText.getStyle().setColor(TextFormatting.DARK_AQUA);
+
+    player.sendMessage(componentText);
+  }
+
+  public static TextFormatting getLevelColor(int level) {
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/CommonProxy.java]: No.

File: src/main/java/slimeknights/toolleveling/Tooltips.java [hunk 9]
```diff
@@ -1,99 +1,48 @@
...
+final class Tooltips {
   private Tooltips() {
   } // non-instantiable
 
   public static void addTooltips(ItemStack itemStack, List<String> tooltips) {
     NBTTagCompound tag = TinkerUtil.getModifierTag(itemStack, TinkerToolLeveling.modToolLeveling.getModifierIdentifier());
     if(!tag.hasNoTags()) {
       ToolLevelNBT data = new ToolLevelNBT(tag);
-      if(Config.canLevelUp(data.level)) {
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/Tooltips.java]: Why would you possibly pull 2 lines into a single line? Even worse, passing a boolean so that a function does A or B is bad, you use 2 functions and call the one that does the logic you need depeding on your condition, not the other way around. Revert to as it was before. It also is more readable that way.

File: src/main/java/slimeknights/toolleveling/ToolLevelNBT.java [hunk 10]
```diff
@@ -1,16 +1,14 @@
...
 
-import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
 
 import slimeknights.tconstruct.library.modifiers.ModifierNBT;
 
-public class ToolLevelNBT extends ModifierNBT {
-
-  private static final String TAG_XP = "xp";
+class ToolLevelNBT extends ModifierNBT {
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/ToolLevelNBT.java]: You don't need to make all classes package protected. Just because we don't use them in another package doesn't mean that somebody else wont use them at some point.

File: src/main/java/slimeknights/toolleveling/config/Config.java [hunk 11]
```diff
@@ -1,37 +1,138 @@
...
 
-  public static int getBaseXpForTool(Item item) {
-    ConfigFile.ToolXP toolXP = INSTANCE.configFile.toolxp;
-    return toolXP.baseXpForTool.getOrDefault(item, toolXP.defaultBaseXP);
+    syncConfig();
   }
 
-  public static float getLevelMultiplier() {
-    return INSTANCE.configFile.toolxp.levelMultiplier;
+  public static void syncConfig() {
```
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: Fix in mantle, if it's broken.

File: src/main/resources/assets/tinkertoolleveling/lang/en_us.lang [hunk 12]
```diff
@@ -1,31 +1,7 @@
...
```
Comment (bonii-xx) [inline src/main/resources/assets/tinkertoolleveling/lang/en_us.lang]: No.

File: src/main/java/slimeknights/toolleveling/CommonProxy.java (hunk 8 above)
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/CommonProxy.java]: The color is supposed to change every level.

File: src/main/resources/assets/tinkertoolleveling/lang/en_us.lang [hunk 13]
```diff
@@ -1,31 +1,7 @@
...
-tooltip.level.9=Legendary
-tooltip.level.10=Godlike
-tooltip.level.11=Awesome
-tooltip.level.19=MoxieGrrl
-tooltip.level.42=boni
-tooltip.level.66=Jadedcat
-tooltip.level.99=Hacker
+message.levelup=Your %s has reached level %s
 
-message.levelup.2=You begin to feel comfortable handling the %s
```
Comment (bonii-xx) [inline src/main/resources/assets/tinkertoolleveling/lang/en_us.lang]: No.

**Comment by bonii-xx** [comment]
>The only big change is the configuration loading. The AbstractConfig.load() would not work (MC crash) so I decided to fix this with "copying" some code from Tinkers Construct istead of fix Mantle to only commit in one place.

No, proper fix is required. Also judging from your code it only broke because you removed all the @mod metadata.

File: build.gradle [hunk 14]
```diff
@@ -3,193 +3,195 @@
  * Master build file
  */
 
+import org.codehaus.groovy.runtime.GStringImpl
```
Comment (DerBaM) [inline build.gradle]: Personelly I like to remove IDE warnings/hints.

File: build.gradle [hunk 15]
```diff
@@ -3,193 +3,195 @@
...
+
 buildscript {
-    repositories {
```
Comment (DerBaM) [inline build.gradle]: Formating change was a seperate push. But this is first time that I use PR. If it make even as seperate push as PR problems I revert it and can it make later as seperate PR. But the much floating/trailing an non formating inner double whitespaces and mix up of tabs and whitespaces should be fixed. I have whitespaces allways activated in my IDE and other editors so I see it. Also make prper formating the source more readable, in my opinion.

File: src/main/java/slimeknights/toolleveling/ModToolLeveling.java (hunk 5 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/ModToolLeveling.java]: There wasn't an emty line between those ifs and that means, that them belong together. And in my opinion it's much more readable now, because you see instand that only one of these conditions must be true.

File: src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java (hunk 7 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java]: Which dependency you mean? I had the Metadata included with dependencies to mantle/tconstruct and it don't work.

File: src/main/java/slimeknights/toolleveling/CommonProxy.java (hunk 8 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/CommonProxy.java]: The code was
`float hue = (0.277777f * level);
hue = hue - (int) hue;`
that means that:
lvl  1 - clumsy:  0.277777
lvl 13 - clumsy+: 0,611101
and yes it change every level, but even the level witch the same name had not the same color (or the have all ?143? cycles).

File: src/main/java/slimeknights/toolleveling/Tooltips.java (hunk 9 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/Tooltips.java]: Because to make a one liner: Level 3   XP 3/400
But I wasn't raelly happy about my solution.

File: src/main/java/slimeknights/toolleveling/ToolLevelNBT.java (hunk 10 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/ToolLevelNBT.java]: "Just because we don't use them in another package" thats a bad reason.

File: src/main/java/slimeknights/toolleveling/config/Config.java (hunk 11 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: It's now use the style of TC so question back: why TC don't use mantle and fix it but has it's own implementation?

**Comment by DerBaM** [comment]
" because you removed all the @mod metadata" as a seperate LATER commit, right.

File: build.gradle (hunk 15 above)
Comment (bonii-xx) [inline build.gradle]: https://blog.jooq.org/2015/03/16/the-10-things-everyone-does-wrong-when-committing-pull-requests/

File: src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java (hunk 7 above)
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java]: The one in the @Mod annotation, that you removed. And yes it does work. That's what the whole "after" part in the dependency is about.

File: src/main/java/slimeknights/toolleveling/CommonProxy.java (hunk 8 above)
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/CommonProxy.java]: Same name doesn't require same color. And no, your code only changes color after the first, the 3rd, the 7th etc. level.

File: build.gradle (hunk 15 above)
Comment (DerBaM) [inline build.gradle]: I'll look at it later and try to make it better.

File: src/main/java/slimeknights/toolleveling/config/Config.java (hunk 11 above)
Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: That is weird indeed. Might be it never got pulled back into Mantle.. or that it requires mod-specific things. What I do know however is that it used to work for TTL, and if it doesn't now then the question is why.

File: src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java (hunk 7 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/TinkerToolLeveling.java]: No, that @mod stuff wasn't removed on the first commit/push and it doesn't work. I've tested it several times and only after I put this it worked. And I included the missing @mod depeencies in my working brunch (which fixed a few positions you mentioned). But first I've finished the 'clumsy' tooltip stuff and then I look into your mentioned 10 things befor I push it.

**Comment by bonii-xx** [comment]
I didn't have time to review it earlier, simple as that. Yes I'm harsh, after all I need to keep maintaining it afterwards too. You wanted to do a port, however most of what you did is not even relevant to a port to 1.12 and made changes on a whim. I'm not saying that the styling is "correct" as is, however formatting does not belong into a PR for a port. If anything it needs a separate PR.

I've put it somewhere above in a comment, but I'm gonna put it here again: https://blog.jooq.org/2015/03/16/the-10-things-everyone-does-wrong-when-committing-pull-requests/

Concentrate on the actual port, and keep the changes only to things that matter to the port. Afterwards there's still time to look at everything else. Every line of formatting changed etc. just adds clutter to the PR.

File: src/main/java/slimeknights/toolleveling/CommonProxy.java (hunk 8 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/CommonProxy.java]: Thats right, I've included 10 quality steps. Currently I included the possibility to change the steps and color in the config file.

**Comment by DerBaM** [comment]
I'll look at your mentioned blog later and do my best.

File: src/main/java/slimeknights/toolleveling/config/Config.java (hunk 11 above)
Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: While I change the PR to be step-by-step I can give you the error message (in the file). So you can tell if I make here something wrong or it is truelly a failure of Mantle (I'll wait till I proceed). The game starts anyway but the configuration wouldn't load so that you have default values.

[mantle_config_read_error.txt](https://github.com/SlimeKnights/TinkersToolLeveling/files/1222661/mantle_config_read_error.txt)

Comment (bonii-xx) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: According to that the config file it tries to read is formatted incorrectly. Try deleting it, and if it persists I suggest we simply deactivate the config loading for now (aka only default values) and tackle that separately.

Comment (DerBaM) [inline src/main/java/slimeknights/toolleveling/config/Config.java]: For the above error message but a running not config-loaded game it needs a config file with at least one value. This musn't be one which are needed. In any other cases (no values, emty, no file) the game crash with an other message.
[crash-2017-08-14_18.52.24-client.txt](https://github.com/SlimeKnights/TinkersToolLeveling/files/1222898/crash-2017-08-14_18.52.24-client.txt)

**Comment by DerBaM** [comment]
bare 1.12 version

It's tested (levelup of tools/weapons, messages, ding sound; on client and server) but configuration is disabled. There diferent error messages depends on if a config file exists or not. Wrote a little more about this above: show outdated Config.java

**Comment by bonii-xx** [comment]
Why did you close the PR now?

**Comment by DerBaM** [comment]
My failure was, that I don't knew that every push will update the PR. So I had (and do in future) to make seperate branches for PR and personal developement. Sorry for that.

But to contribute on this project the response time is to long for me. That's why I decided to make a fork on which I actually work which is savegamecompatible to this but has it's own lightwight flavor. To not steal you more time and don't confuse other users I closed this and get the repo offline.
